From 42b47dd718d08e780b8a9cd07a4e22e07fdc719d Mon Sep 17 00:00:00 2001
From: stephen hemminger <stephen@networkplumber.org>
Date: Wed, 19 Jul 2017 11:53:18 -0700
Subject: [PATCH 36/98] netvsc: save pointer to parent netvsc_device in channel
 table

Keep back pointer in the per-channel data structure to
avoid any possible RCU related issues when napi poll is
called but netvsc_device is in RCU limbo.

Signed-off-by: Stephen Hemminger <sthemmin@microsoft.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---
 drivers/net/hyperv/hyperv_net.h |    1 +
 drivers/net/hyperv/netvsc.c     |    3 ++-
 2 files changed, 3 insertions(+), 1 deletion(-)

--- a/drivers/net/hyperv/hyperv_net.h	2017-11-17 10:19:53.755285541 -0800
+++ b/drivers/net/hyperv/hyperv_net.h	2017-11-17 10:19:53.751285593 -0800
@@ -727,6 +727,7 @@ struct net_device_context {
 /* Per channel data */
 struct netvsc_channel {
 	struct vmbus_channel *channel;
+	struct netvsc_device *net_device;
 	const struct vmpacket_descriptor *desc;
 	struct napi_struct napi;
 	struct multi_send_data msd;
--- a/drivers/net/hyperv/netvsc.c	2017-11-17 10:19:53.755285541 -0800
+++ b/drivers/net/hyperv/netvsc.c	2017-11-17 10:19:53.751285593 -0800
@@ -1226,11 +1226,11 @@ int netvsc_poll(struct napi_struct *napi
 {
 	struct netvsc_channel *nvchan
 		= container_of(napi, struct netvsc_channel, napi);
+	struct netvsc_device *net_device = nvchan->net_device;
 	struct vmbus_channel *channel = nvchan->channel;
 	struct hv_device *device = netvsc_channel_to_device(channel);
 	u16 q_idx = channel->offermsg.offer.sub_channel_index;
 	struct net_device *ndev = hv_get_drvdata(device);
-	struct netvsc_device *net_device = net_device_to_netvsc_device(ndev);
 	int work_done = 0;
 
 	/* If starting a new interval */
@@ -1314,6 +1314,7 @@ struct netvsc_device *netvsc_device_add(
 		struct netvsc_channel *nvchan = &net_device->chan_table[i];
 
 		nvchan->channel = device->channel;
+		nvchan->net_device = net_device;
 		u64_stats_init(&nvchan->tx_stats.syncp);
 		u64_stats_init(&nvchan->rx_stats.syncp);
 	}
